File: simulator/python_implementations/BST.py
from DictionaryImplementor import *

class BSTNode(object):
    def __init__(self,val, left, right):
        self.val = val
        self.left = left
        self.right = right

class BST(DictionaryImplementor):

    @staticmethod
    def search(root,x):
        if not root:
            return False

        if root.val == x :
            return True
        elif  x < root.val:
            return BST.search(root.left, x)
        else:
            return BST.search(root.right,x)

    @staticmethod
    def insert(root,x):

        toInsert = BSTNode(x,None,None)

        if not root:
            root = toInsert
        else:
            BST.__insertNode(root, toInsert)

        return root

    @staticmethod
    def __insertNode(root,toInsert):

        if toInsert.val < root.val:
            if not root.left:
                root.left = toInsert
                return
            BST.__insertNode(root.left, toInsert)
        else:
            if not root.right:
                root.right = toInsert
                return
            BST.__insertNode(root.right,toInsert)



    @staticmethod
    def delete(root,x):
        if not root:
            return None

        if root.val == x:
            if not root.right and not root.left:
                return None
            elif root.right and not root.left:
                return root.right
            elif not root.right and root.left:
                return root.left
            else:
                min = BST.__findMin(root)
                root.val = min.val
                root.left = BST.delete(root.left,min.val)
        else:
            root.left = BST.delete(root.left,x)
            root.right = BST.delete(root.right,x)
        return root


    @staticmethod
    def __findMin(root):
        if root.left:
            return BST.__findMin(root.left)
        else:
            return root

    @staticmethod
    def sort():
        pass

    @staticmethod
    def toStringInOrder(root):
        return BST.__inorderAppend(root)

    @staticmethod
    def __inorderAppend(root):
        s = ""
        if not root:
            return s
        else:
            s+=BST.__inorderAppend(root.left)
            s+=" "+str(root.val)
            s+=BST.__inorderAppend(root.right)
            return s


    @staticmethod
    def insertInOrder(root,a):
        for val in a:
            BST.__insert(root,val)

        return root

File: src/cpp/BST/BST.cpp
#include "BST.h"


void BST::insert(int to_insert)
{
  if(root == NULL)
  {
    this->root = new BSTNode(NULL,NULL,to_insert);
  }
  else
    insert(to_insert,*root);
}


void BST::insert(int to_insert, BSTNode& node)
{
  if(&node == NULL)
     return;
  else if (node.val == to_insert)
    return;
  else if (node.val > to_insert)
    if (node.left == NULL)
    {
      BSTNode* new_node = new BSTNode(NULL,NULL,to_insert);
      node.left = new_node;
    }
    else
      insert(to_insert, *node.left);
  else
  {
    if (node.right == NULL)
    {
      BSTNode* new_node = new BSTNode(NULL,NULL,to_insert);
      node.right = new_node;
    }
    else
      insert(to_insert, *node.right);
  }
}


bool BST::search(int to_find)
{
  return search(to_find,*root);
}

bool BST::search(int to_find, BSTNode& node)
{
  if(&node == NULL)
  {
    return false;
  }
  else if (node.val == to_find)
  {
    return true;
  }
  else
  {
    if (( to_find < node.val) && (node.left  != NULL))
      return search(to_find, *node.left);
    else if (node.right  != NULL)
      return search(to_find, *node.right);
  }
  return false;
}


void BST::del(int to_delete)
{
   if (root != NULL)
   {
     // deal with case of root here
     if( root->val == to_delete)
      del_parent(root,NULL);
     else
      del_search(to_delete,*root);
   }

}

void BST::del_search(int to_delete, BSTNode& node)
{

  if(&node == NULL)
    return;
  else if (node.left != NULL)
  {
    if (node.left->val == to_delete)
    {
      del_parent(node.left,&node);
    }
    else
      del_search(to_delete, *node.left);
  }
  else if (node.right != NULL)
  {
    if (node.right->val == to_delete)
    {
      del_parent(node.right,&node);
    }
    else
      del_search(to_delete, *node.right);
  }

}

BSTNode* BST::find_min(BSTNode* node)
{
  if(node->left !=  NULL)
  {
    return find_min(node->left);
  }
  return node;
}


void BST::del_parent(BSTNode* to_delete, BSTNode* parent)
{
  if ((to_delete->right ==  NULL) && (to_delete->left == NULL))
  {
    if (parent == NULL)
    {
      this->root =  NULL;
    }
    else
    {
      if (to_delete == parent->right)
      {
        parent->right =  NULL;
      }
      else
      {
        parent->left =  NULL;
      }
    }
  }
    else if ((to_delete->right != NULL) && (to_delete->left == NULL))
    {
      if (parent == NULL)
      {
        this->root =  to_delete->right;
      }
      else
      {
        if (to_delete == parent->right)
        {
          parent->right =  to_delete->right;
        }
        else
        {
          parent->left =  to_delete->right;
        }
      }
    }
      else if ((to_delete->right == NULL) && (to_delete->left != NULL))
      {
        if (parent == NULL)
        {
          this->root =  to_delete->left;
        }
        else
        {
          if (to_delete == parent->right)
          {
            parent->right =  to_delete->left;
          }
          else
          {
            parent->left =  to_delete->left;
          }
        }
      }
     else
     {
          BSTNode* left_min = find_min(to_delete->left);
          to_delete->val = left_min->val;
          del_search(left_min->val, *to_delete);
      }

}




string BST::to_string_in_order()
{
  string ret;
  ret = to_string_in_order_traverse(ret,*root);
  return ret;
}


string BST::to_string_in_order_traverse(string& ret,BSTNode& node)
{
  if(&node != NULL)
  {
    ret = to_string_in_order_traverse(ret,*node.left);
    if(ret.length() == 0)
    {
      ret=ret+to_string(node.val);
    }
    else
    {
      ret=ret+" "+to_string(node.val);
    }
    ret = to_string_in_order_traverse(ret,*node.right);
  }

  return ret;
}
